sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,538 @@
+represented by the following struct:
+```
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+List of imported symbols is represented as a single property set with name
+`SYCL/imported symbols` recorded in the `Name` field of property set.
+Each property in this set holds name of the particular imported symbol recorded
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `SYCL/exported symbols`.
+
+### DPC++ runtime changes
+
+DPC++ RT performs *device images collection* task by grouping all device
+images required to execute a kernel based on the list of exports/imports and
+links them together using PI API.
+This native device image is then added to the cache to avoid symbol resolution,
+compilation, and linking for any future attempts to invoke kernels from this
+device image.
+
+#### DPC++ runtime plugin interface (PI) changes
+
+Before creating a program the function `piextDeviceSelectBinary` is used to
+choose the most appropriate device image. It is possible that not all backends
+have possibility to link device images of particular format at run-time. So,
+in presence of dynamic linking the `piextDeviceSelectBinary` function should be
+extended, so it chooses the appropriate device image
+based on additional attributes of device images, such as:
+- List of imports (if present)
+- Device image format
+- Runtime linking support in corresponding backend
+
+Example: the backend doesn't have support of native binaries linking at
+run-time but linking of SPIR-V device images is supported,
+the AOT-compiled device image with required kernel have imports
+information attached which effectively means that this device image needs runtime
+linking, but since native binaries linking is not supported, the image with
+SPIR-V format will be chosen (or an error emitted if there is no other device
+images).
+To link several device images together `piProgramLink` API will be used.
+Depending on concrete plugin implementation and set of device image formats that
+can be linked at run-time, `piProgramLink` API may receive device images in
+different states as inputs (including SPIR-V and native code) with a limitation
+that all inputs should have the same format.
+
+##### Support of runtime linking in backends
+
+- The initial design will support dynamic linking of device code in SPIR-V
+  format on OpenCL backend:
+  - OpenCL plugin will use the existing OpenCL `clLinkProgram()` API to online
+  link the SPIR-V modules together.
+- The initial design will support dynamic linking of device code in native code
+  format on the Level Zero backend:
+  - L0 plugin will use the existing Level Zero `zeModuleDynamicLink()` API to do
+  the linking.
```

Comment:
It seems like this uneven backend support will make it very difficult to use this new feature.  If the user wants to run on the CPU device (OpeCL backend), they **must** compile in AOT mode.  However, if the user wants to run on a GPU (Level Zero by default), they **must** compile w/o AOT mode.  There's no single compilation mode that will work for all devices.

I think it would be a big improvement if we could extend Level Zero to have the ability to online link SPIR-V modules.  This way, at least JIT mode would support all device types.

Reply:
Ok, that makes sense. I mentioned that new L0 API for SPIR-V linking is required.

Reply:
> If the user wants to run on the CPU device (OpeCL backend), they must compile in AOT mode. However, if the user wants to run on a GPU (Level Zero by default), they must compile w/o AOT mode. There's no single compilation mode that will work for all devices.

It actually the opposite. However, looking at [Module section](https://spec.oneapi.com/level-zero/latest/core/api.html#module) of Level Zero specification, I can't find any limitations about `zeModuleDynamicLink`, so I would assume that it is also capable of linking `SPIR-V` images as well.

Reply:
> However, looking at Module section of Level Zero specification, I can't find any limitations about zeModuleDynamicLink, so I would assume that it is also capable of linking SPIR-V images as well.

In Level Zero, a "module" is always native device code.  There's no such thing as a SPIR-V format module.  Therefore, zeModuleDynamicLink only links native device code, not SPIR-V.

Reply:
> > However, looking at Module section of Level Zero specification, I can't find any limitations about zeModuleDynamicLink, so I would assume that it is also capable of linking SPIR-V images as well.
>
> In Level Zero, a "module" is always native device code. There's no such thing as a SPIR-V format module. Therefore, zeModuleDynamicLink only links native device code, not SPIR-V.

Fair enough, but since we should be able to create such modules from SPIR-V, to DPC++ RT layer it will look like SPIR-V can be linked as well. Unless I'm missing something from Level Zero spec

Reply:
> Fair enough, but since we should be able to create such modules from SPIR-V, to DPC++ RT layer it will look like SPIR-V can be linked as well. Unless I'm missing something from Level Zero spec

Yes, this is possible, but I think it is not the best solution.  Relying on zeModuleDynamicLink is less efficient for two reasons: A) you lose any possibility of cross-module optimizations because those optimizations can only be done when the input is an IL like SPIR-V, and B) the "in-place" patching semantics of zeModuleDynamicLink mean that you sometimes need to make a copy of the input modules (this is because a second link operation might undo the effect of a previous link operation).

Since other DPC++ features need the SPIR-V linking API anyways, it seems better to make use of it with this shared library feature too in the case when the shared libraries contain SPIR-V.

Reply:
> In Level Zero, a "module" is always native device code. There's no such thing as a SPIR-V format module. Therefore, zeModuleDynamicLink only links native device code, not SPIR-V.

From pure logic standpoint, not necessarily, actually. L0 linker could have SPIRV inputs, link them, then compile to produce native output, w/o ability to return the intermediate link result. Since with dynamic linking we never do tree- or graph-like hierarchical linking (it is always flat - N inputs and 1 output), that interface could be enough. But that'd be really inflexible, and either existing L0 link API should be extended to allow SPIRV inputs/SPIRV output or a new API added.

But, I think I wrote this before, there should be nothing device-specific in SPIR-V linkage process, and it would be more flexible and efficient (from code duplication standpoint) to host SPIRV linkage to SYCL RT level.
- more flexible means there can be runtimes which don't support SPIRV linkage but support SPIRV inputs (like LevelZero :) )
- efficient means all BEs don't need to repeat the spirv-link implementation.

If BE has SPIRV link capability, SYCL RT might decide to use it instead of the generic SPIRV linker it (SYCL RT) would have built-in.

Reply:
> Since with dynamic linking we never do tree- or graph-like hierarchical linking (it is always flat - N inputs and 1 output), that interface could be enough.

The Level Zero `zeModuleDynamicLink` API does not have any outputs.  It does an "in-place" link, where the inputs are modified to perform the linking operation.  I guess your larger point is that the existing API could be used to do SPIR-v linking.  I'm not sure I agree with this, but the point is moot.  The existing implementation does not work this way.  Level Zero modules are native code, and `zeModuleDynamicLink` does an in-place link of the native code.

> But, I think I wrote this before, there should be nothing device-specific in SPIR-V linkage process, and it would be more flexible and efficient (from code duplication standpoint) to hoist SPIRV linkage to SYCL RT level.

Yes, I think we had this conversation already.  :-)  I agree that adding a SPIR-V linking phase into the DPC++ gives you more options.  However, I don't agree about the efficiency argument.  There are two types of efficiency:

* Runtime efficiency.  Here I think it would be less efficient if DPC++ links the SPIR-V because that would be a separate pass over the SPIR-V module.  DPC++ would read SPIR-V and then write another SPIR-V.  The backend would then read the second SPIR-V.  Compare this with a design where the backend reads the input SPIR-V only once and has a combined pass that does both linking and native code generation.

* Engineering efficiency.  I think this is what you are arguing.  I don't agree, though.  The DPC++ runtime does not currently have any dependency on SPIR-V, so this would be a new dependency if we add it to the runtime.  I think that makes it less efficient in an engineering sense.  If the backend does SPIR-V linking, then it may seem like this logic needs to be duplicated in both OpenCL and Level Zero, but I think this is not really the case.  Both of those backends use IGC as the JIT compiler, so the logic is in just one place.  Moreover, IGC must already know how to do SPIR-V linking because this is how OpenCL works today.  Therefore, we just want to expose that existing IGC capability through the Level Zero API.

Reply:
> The Level Zero zeModuleDynamicLink API does not have any outputs.
...
> The existing implementation does not work this way. Level Zero modules are native code, and zeModuleDynamicLink does an in-place link of the native code.

Yes, I understand this. But that does not invalidate the statements I made, as far as I can see. In-place linking is inconvenient, but the linker API still has the output, represented in an exotic way though.

> Compare this with a design where the backend reads the input SPIR-V only once and has a combined pass that does both linking and native code generation.

In practice, I strongly suspect the additional time to parse SPIRV would be dwarfed by the JIT time. Besides, I proposed: "If BE has SPIRV link capability, SYCL RT might decide to use it instead of the generic SPIRV linker it (SYCL RT) would have built-in.". So I don't think what I suggested would lead to any performance degradation if designed properly. But there are  benefits
 - SPIRV dynamic linkage can still be supported for BEs w/o SPIRV linkage capabilities 
 - The SPIRV linkage could be exposed to the user giving ultimate flexibility for RT code management

> The DPC++ runtime does not currently have any dependency on SPIR-V, so this would be a new dependency if we add it to the runtime.

I disagree. DPC++ runtime does have SPIRV dependence. E.g. before creating a program it checks binary format, and redirects to createProgramWithIL (PI equivalent of it) if it is SPIRV. DPC++ as a whole relies on SPIRV as portable device code format and provides SPIRV tools, so: I neither agree with this:

> so this would be a new dependency if we add it to the runtime. I think that makes it less efficient in an engineering sense.

> If the backend does SPIR-V linking, then it may seem like this logic needs to be duplicated in both OpenCL and Level Zero, but I think this is not really the case.

the duplication will be there in case of multiple independent BEs who want to enjoy dynamic linking - from vendor X and vendor Y. It needs to be there even today, I believe - consider FPGA and CPU BEs.

Reply:
One more important factor in having SPIRV linkage exposed independently of BE - it does not depend on particular BE :) On conceptual level, why one would need context or other BE binding to be able to link SPIRV in the host part of SYCL app?